lib/crypto/src/arithmetic/uint.rs

```diff
@@ -0,0 +1,985 @@
+use alloc::string::String;
+use core::{
+    borrow::Borrow,
+    cmp::Ordering,
+    fmt::{Debug, Display, Result, UpperHex, Write},
+    ops::{
+        BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXor, BitXorAssign, Not,
+        Shl, ShlAssign, Shr, ShrAssign,
+    },
+};
+
+use num_traits::ConstZero;
+use zeroize::Zeroize;
+
+use crate::{
+    arithmetic::{
+        limb,
+        limb::{Limb, Limbs},
+        BigInteger,
+    },
+    bits::BitIteratorBE,
+    ct_for, ct_for_unroll6,
+};
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Zeroize)]
+pub struct Uint<const N: usize> {
+    pub(crate) limbs: Limbs<N>,
+}
+
+impl<const N: usize> Default for Uint<N> {
+    fn default() -> Self {
+        Self { limbs: [Limb::ZERO; N] }
+    }
+}
+
+/// Declare [`Uint`] types for different bit sizes.
+macro_rules! declare_num {
+    ($num:ident, $bits:expr) => {
+        #[doc = "Unsigned integer with "]
+        #[doc = stringify!($bits)]
+        #[doc = "bits size."]
+        pub type $num = $crate::arithmetic::uint::Uint<
+            { usize::div_ceil($bits, $crate::arithmetic::Limb::BITS as usize) },
+        >;
+    };
+}
+
+declare_num!(U64, 64);
+declare_num!(U128, 128);
+declare_num!(U192, 192);
+declare_num!(U256, 256);
+declare_num!(U384, 384);
+declare_num!(U448, 448);
+declare_num!(U512, 512);
+declare_num!(U576, 576);
+declare_num!(U640, 640);
+declare_num!(U704, 704);
+declare_num!(U768, 768);
+declare_num!(U832, 832);
+
+impl<const N: usize> Uint<N> {
+    #[must_use]
+    pub const fn new(value: [Limb; N]) -> Self {
+        Self { limbs: value }
+    }
+
+    #[must_use]
+    pub const fn as_limbs(&self) -> &[Limb; N] {
+        &self.limbs
+    }
+
+    #[doc(hidden)]
+    #[inline]
+    #[must_use]
+    pub const fn ct_is_odd(&self) -> bool {
+        self.limbs[0] & 1 == 1
+    }
+
+    #[doc(hidden)]
+    #[inline]
+    pub const fn ct_is_even(&self) -> bool {
+        !self.ct_is_odd()
```

Comment:
Same here for optimisation; 
You can avoid:
- `ct_is_odd()` invocation
- Negation (`!`)

By only having `self.limbs[0] & 1 == 0`